File: Top_Interview_Questions/Medium/152. Maximum Product Subarray/fail_solution.py
from typing import List
import sys


class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        product, max_product = 1, -sys.maxsize
        negative_pos = []
        for i, num in enumerate(nums):
            if num < 0:
                negative_pos.append(i)
            product *= num
        if len(negative_pos) % 2 == 0:
            return product
        num_negtaive = len(negative_pos)
        start = -1
        for i in range(num_negtaive):
            if i + num_negtaive - 1 > num_negtaive:
                break
            product = 1
            end = negative_pos[i + num_negtaive - 1] if i + num_negtaive - 1 < num_negtaive else len(nums)
            for j in range(start + 1, end):
                product *= nums[j]
            max_product = max(max_product, product)
            start = negative_pos[i]
        return max_product

File: Top_Interview_Questions/Medium/152. Maximum Product Subarray/fail_solution.cpp
#include <vector>
using namespace std;

/**
 * 想法: 分成奇數個負數與偶數個負數來討論：
 *      1. 偶數個負數：把整個 nums 都乘起來就是答案
 *      2. 奇數個負數：設有 n 個負數，討論 n-1 個(偶數個)負數的組合
 *         例如：[3,-2,4,-5,-10]
 *          -> -2,-5,-10 三個負數 -> 討論 2 個負數的情況
 *          -> (-2,-5): 3*-2*4*-5=120
 *          -> (-5,-10): 4*-5*-10=200
 *          -> 組合 2 有最大乘積
 * (失敗，忘記考慮 nums 會有 0 存在)
 * 時間: O(n), 空間: O(n)
 */

class Solution
{
  public:
    int maxProduct(vector<int> &nums)
    {
        int product = 1, maxProduct = INT_MIN;
        vector<int> negativePos;
        for (int i = 0; i < nums.size(); i++)
        {
            if (nums[i] < 0)
            {
                negativePos.push_back(i);
            }
            product *= nums[i];
        }
        if (negativePos.size() % 2 == 0)
        {
            return product;
        }
        int numNegative = negativePos.size();
        int start = -1;
        for (int i = 0; i + numNegative - 1 <= numNegative; i++)
        {
            product = 1;
            int end = i + numNegative - 1 < numNegative ? negativePos[i + numNegative - 1] : nums.size();
            for (int j = start + 1; j < end; j++)
            {
                product *= nums[j];
            }
            maxProduct = max(maxProduct, product);
            start = negativePos[i];
        }
        return maxProduct;
    }
};